repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py

```diff
@@ -0,0 +1,339 @@
+      - directives - list of strings
+      - new-directive - string
+      - file - string
+  
+      returns boolean
+      """
+
+      result = False
+
+      for directive in directives:
+        # regexp and new string differs if we replace directive or value - directive needs to have at least one
+        # white character at the end, but value needs to have one white character at start and zero or N white
+        # characters at the end + new line character
+        if directive in changed_directives:
+          directive_regexp = directive + '\s'
+          new_directive_string = new_directive + ' '
+        elif directive in changed_values:
+          directive_regexp = '\s' + directive + '\s*\n'
+          new_directive_string = ' ' + new_directive + '\n'
+        pattern = compile(directive_regexp)
+
+        try:
+          with open(file, 'r') as conf:
+            content = conf.read()
+        except IOError:
+          return result
+
+        new_content = sub(pattern, new_directive_string, content)
+        if new_directive == 'ServerKeychain':
+          # ServerKeyChain migration needs more attention - like moving the files into specific directory
+          found_values = self.find_values(new_directive, new_content)
+          for value in found_values:
+            new_content = self.migrate_serverkeychain_path(value, new_directive, new_content)
+
+        try:
+          with open(file, 'w') as conf:
+            conf.write(new_content)
+          result = True
+        except IOError:
+          return result
+
+      return result
+
+    def migrate_includes(self, include_files):
+      """
+      concatenate_file
+  
+      parameters:
+      - include_files - list of strings
+  
+      Returns boolean
+      """
+
+      try:
+        with open(include_files[0], 'r') as cupsd_conf:
+          cupsd_contents = cupsd_conf.read()
+      except IOError:
+        return False
+        
+      # concatenation of files
+      for conf in include_files[1:]:
+        if os.path.exists(conf) is True:
+          try:
+            with open(conf, 'r') as opened_conf:
+              includefile_contents = opened_conf.read()
+          except IOError:
+            return False
+          try:
+            with open(include_files[0], 'w') as cupsd_conf:
+              cupsd_contents = cupsd_contents + includefile_contents
+              cupsd_conf.write(cupsd_contents)
+          except IOError:
+            return False
+        else:
+          return False
+
+      # remove Include directives from cupsd.conf
+      found_values = self.find_values('Include', cupsd_contents)
+      for value in found_values:
+        cupsd_contents = self.remove_string('Include', value, cupsd_contents)
+
+      try:
+        with open(include_files[0], 'w') as cupsd_conf:
+          cupsd_conf.write(cupsd_contents)
+        return True
+      except IOError:
+        return False
+
+    def move(self, directives, source, target):
```

Comment:
ditto about returning meaningful errors and doing all checks before properly executing something.

Reply:
Done.